packages/cli/src/cmds/preview/preview.ts

```diff
@@ -145,12 +151,14 @@ export default class Preview extends YargsCommand {
       includeBot && skipNgrok ? botRoot : undefined
     );
     if (result.isErr()) {
+      await this.terminateTasks();
       return result;
     }
 
     /* === check ports === */
     const portsInUse = await commonUtils.getPortsInUse(includeFrontend, includeBackend, includeBot);
     if (portsInUse.length > 0) {
+      await this.terminateTasks();
```

Comment:
Call `terminateTasks` via process.on("exit") instead of everywhere...

Reply:
Tried this way. One issue is that the exit callback logic must be synchronous: https://nodejs.org/dist/latest-v14.x/docs/api/process.html#process_event_exit, but the tree-kill operation is synchronous. So this way cannot terminate the processes as expected.